### PR TITLE
Add example Spring Events renderer drawing a sequence diagram

### PR DESCRIPTION
**Pull request by wandernauta**
The renderer takes JSON on standard input and generates Asciidoc with PlantUML on standard output. When fed the Spring Events example (#86), the rendered output looks like this:

2023-11-19-185922_785x639_scrot

### CODE REVIEW

**Comment by wandernauta** [comment]
#73 is related.

File: ldj-spring-renderer-example/src/test/java/com/infosupport/ldoc/springexample/SpringEventRendererTest.java
Comment (GuusVink) [inline ldj-spring-renderer-example/src/test/java/com/infosupport/ldoc/springexample/SpringEventRendererTest.java:28]: Why this test?

File: ldj-spring-renderer-example/src/main/java/com/infosupport/ldoc/springexample/SpringEventRenderer.java
Comment (GuusVink) [inline ldj-spring-renderer-example/src/main/java/com/infosupport/ldoc/springexample/SpringEventRenderer.java:123]: I would say performance is not an issue. The only thing is the readability for people that try to follow the code (considering the exemplifying nature of this renderer)

**Review by GuusVink — COMMENTED** [review]
In general: good work.

The only thing I have doubts about are its use as serving as an easy to understand example of how a renderer can be created. I had to spend quite some time trying to follow the structure of the code (mainly the underlying model that is used along with the naming conventions).

For right now, this is fine, however, I image this example would change once a support library is present (at least the minimal C# equivalent). At that stage (and after the AMPQ renderer has been made) I hope that the renderer can be made somewhat simpler/more understandable.

My only question is whether it would be best practice to merge this branch into main right away. Or keep it in this branch, wait for (minimal) support library, and then merge into main. 
I am wondering what you would think is most appropriate. In case of the former, I am happy to approve this PR, otherwise, we can postpone.

File: ldj-spring-renderer-example/src/test/java/com/infosupport/ldoc/springexample/SpringEventRendererTest.java
Comment (wandernauta) [reply ldj-spring-renderer-example/src/test/java/com/infosupport/ldoc/springexample/SpringEventRendererTest.java]: Ah, I had it to check that JUnit was working, it doesn't serve a purpose anymore. Removed.

**Comment by wandernauta** [comment]
Thanks for the review!

> The only thing I have doubts about are its use as serving as an easy to understand example of how a renderer can be created. I had to spend quite some time trying to follow the structure of the code (mainly the underlying model that is used along with the naming conventions).

This for sure is a good thing to fix now, before it is merged - the idea is indeed that this thing should be easy to follow and understand, and therefore useful as a base. Basically there _should_ be nothing exotic in there; just Jackson Databind, just `for`s and `if`s over `List`s, no stream API and so on. If you do feel there is adventurous logic in there, definitely add a comment; if you know a simpler algorithm to do the same thing, likewise.

For what it's worth, the names come from:

| Name | Source |
| ---- | ------ |
| Listen, listener | Spring |
| Publish, publisher | Spring |
| Interaction | PitstopDocumentationRenderer example |
| Participant | PlantUML, which takes it from UML |

...but obviously if you think I should have taken the names from somewhere else, also add a comment and I'll rename things.

> I imagine this example would change once a support library is present (at least the minimal C# equivalent). At that stage (and after the AMQP renderer has been made) I hope that the renderer can be made somewhat simpler/more understandable.

We'll probably discuss this further, but if that support library is a general-purpose PlantUML-printing library, I feel it would not save a lot of lines of code in this example. Really only four lines have some PlantUML-specific things in them (line 114, line 118, line 131 and line 133) and my prediction is that if we used some library for this, we would replace those four lines with four different lines.

> My only question is whether it would be best practice to merge this branch into main right away. 

My preference would go toward fixing any understandability issues we see right now, then merging.

**Comment by GuusVink** [comment]
Btw, I now notice that two of the tests are failing (render, and renderInteractions). While I do see that the expected and actual values seem to be the same. To verify, do these tests also fail on your machine?

**Comment by wandernauta** [comment]
> Btw, I now notice that two of the tests are failing (render, and renderInteractions).

Do you happen to have a Windows machine? Are they failing because of whitespace differences?

**Comment by GuusVink** [comment]
The IntelliJ editor does not see any differences. So while I suspect whitespaces/line endings are the problem, I can not see this

**Comment by wandernauta** [comment]
I've pushed a change that should make it so that `\n` is always used, regardless of platform. Does that fix the issue?

**Comment by wandernauta** [comment]
> That fixes the issue regarding tests. However, the QA gate fails due to a code smell `should use %n instead of \n`

This is basically a false positive. I don't want a platform-dependent end-of-line character here; I want surgically a newline `\n`, regardless of platform, for consistency both with the `.java` file and the template `.adoc` file.

> What would you consider the best thing to do?

I would suggest merging the PR then asking Ali to mark the suggestion as not useful inside the Sonar UI.

**Comment by GuusVink** [comment]
👍🏼

**Review by GuusVink — APPROVED** [review]
Good for merge, some future work possible for making the sample renderer more understandable. This will be in a separate PR

**Pull request merged**
